### PR TITLE
Support `$rand` query operator

### PR DESCRIPTION
**Pull request by ribaraka**
Refs #258.

$rand operator has been implemented for `update` command. 
For the `find` command, the `$expr` operator must first be implemented.

### CODE REVIEW

File: internal/handlers/common/params.go [hunk 1]
```diff
@@ -57,6 +57,25 @@ func GetOptionalParam[T types.Type](doc *types.Document, key string, defaultValu
 	return res, nil
 }
 
+// GetCompositeParam returns doc's value of composite data type for key or protocol error for missing or invalid parameter.
+func GetCompositeParam(doc *types.Document, key string) (zero any, err error) {
```
Comment (w84thesun) [inline internal/handlers/common/params.go:61]: Let's not use named result parameters as we don't really need them.

File: internal/handlers/common/params.go [hunk 2]
```diff
@@ -57,6 +57,25 @@ func GetOptionalParam[T types.Type](doc *types.Document, key string, defaultValu
...
+	v, err := doc.Get(key)
+	if err != nil {
+		msg := fmt.Sprintf("required parameter %q is missing", key)
+		return zero, NewErrorMsg(ErrBadValue, msg)
```
Comment (w84thesun) [inline internal/handlers/common/params.go:65]:
```suggestion
		return nil, NewErrorMsg(ErrBadValue, fmt.Sprintf("required parameter %q is missing", key))
```

File: internal/handlers/common/params.go [hunk 3]
```diff
@@ -57,6 +57,25 @@ func GetOptionalParam[T types.Type](doc *types.Document, key string, defaultValu
...
+	}
+
+	switch v := v.(type) {
+	case *types.Document:
+		return v, nil
+	case *types.Array:
+		return v, nil
+	default:
+		msg := fmt.Sprintf("composite parameter %q has type %T", key, v)
+		return zero, NewErrorMsg(ErrBadValue, msg)
```
Comment (w84thesun) [inline internal/handlers/common/params.go:75]:
```suggestion
		return nil, NewErrorMsg(ErrBadValue, fmt.Sprintf("composite parameter %q has type %T", key, v))
```

File: internal/handlers/common/params.go (hunk 1 above)
Comment (ribaraka) [reply internal/handlers/common/params.go]: fixed.

File: internal/handlers/common/params.go (hunk 2 above)
Comment (ribaraka) [reply internal/handlers/common/params.go]: fixed

File: internal/handlers/common/params.go (hunk 3 above)
Comment (ribaraka) [reply internal/handlers/common/params.go]: fixed

**Review by w84thesun — APPROVED** [review]

File: internal/handlers/common/update.go [hunk 4]
```diff
@@ -16,16 +16,34 @@ package common
...
 )
 
 // UpdateDocument updates the given document with a series of update operators.
-func UpdateDocument(doc, update *types.Document) error {
-	for _, updateOp := range update.Keys() {
-		updateV := must.NotFail(update.Get(updateOp))
+func UpdateDocument(doc *types.Document, update any) error {
+	var updateDoc *types.Document
+	var isUpdateArray bool
+	for {
```
Comment (seeforschauer) [inline internal/handlers/common/update.go:30]: Sorry what for here is an endless `for`? `update` is not changing.

**Review by seeforschauer — CHANGES_REQUESTED** [review]
looks like endless for is extra

File: internal/handlers/common/update.go (hunk 4 above)
Comment (ribaraka) [reply internal/handlers/common/update.go]: Yep, I overdid it just a bit. :smile: 
Good catch. Thanks! Fixed

**Review by seeforschauer — CHANGES_REQUESTED** [review]
Thank you!
can you please add tests if one wants to update something like
`update := bson.A{bson.D{{"$set", 
bson.D{{"foo", bson.D{{"bar": bson.D{{"baz", bson.M{"$rand": bson.D{}} }} }} }} } }}`

**Comment by ribaraka** [comment]
> Thank you! can you please add tests if one wants to update something like `update := bson.A{bson.D{{"$set", bson.D{{"foo", bson.D{{"bar": bson.D{{"baz", bson.M{"$rand": bson.D{}} }} }} }} } }}`

This is the second time I have met the need for `deepEqual` in a document :smile:

**Comment by seeforschauer** [comment]
> > Thank you! can you please add tests if one wants to update something like `update := bson.A{bson.D{{"$set", bson.D{{"foo", bson.D{{"bar": bson.D{{"baz", bson.M{"$rand": bson.D{}} }} }} }} } }}`
> 
> This is the second time I have met the need for `deepEqual` in a document smile Maybe this time is not exactly the same case as the first one, but something like that might help.
> 
> Added the test, it's failing. Working on fixing, current implementation is not flexible.

Use of `reflect.Deep` is unlikely, preferably write a recursive function.

**Comment by AlekSi** [comment]
This ticket was specifically about `$rand` query operator, not about the aggregation pipeline operator. They are very separate.

**Comment by ribaraka** [comment]
> This ticket was specifically about `$rand` query operator, not about the aggregation pipeline operator. They are very separate.

This comment relates to me? Bcs honestly, I do not understand what do you mean. I follow this https://www.mongodb.com/docs/manual/reference/operator/query/rand/

**Comment by ribaraka** [comment]
> I guess there is no way to implement this issue now.
> 
> For queries, `$rand` can be used only inside `$expr`, and `$expr` requires a full aggregation pipeline support that we don't want to implement now (that's item 21 on our [roadmap](https://github.com/orgs/FerretDB/projects/2/views/1), and we are on item 2 now).
> 
> When used in updates, the array there is the same aggregation pipeline.
> 
> Sorry for confusion.

No worries. Thanks for answers :+1:

**Pull request closed without merging**